File: pixie/instrumentation/spans.py
"""Data model types for pixie instrumentation spans."""

from __future__ import annotations

from dataclasses import dataclass
from datetime import datetime
from typing import Any, Literal  # noqa: UP035

# ── Message content types ─────────────────────────────────────────────────────


@dataclass(frozen=True)
class TextContent:
    """Plain text content part."""

    text: str
    type: Literal["text"] = "text"


@dataclass(frozen=True)
class ImageContent:
    """Image content part (URL or data URI)."""

    url: str  # https:// or data: URI
    detail: str | None = None  # "low" | "high" | "auto" | None
    type: Literal["image"] = "image"


MessageContent = TextContent | ImageContent


# ── Tool types ────────────────────────────────────────────────────────────────


@dataclass(frozen=True)
class ToolCall:
    """Tool invocation requested by the model."""

    name: str
    arguments: dict[str, Any]  # always deserialized, never a raw JSON string
    id: str | None = None


@dataclass(frozen=True)
class ToolDefinition:
    """Tool made available to the model in the request."""

    name: str
    description: str | None = None
    parameters: dict[str, Any] | None = None  # JSON Schema object


# ── Message types ─────────────────────────────────────────────────────────────


@dataclass(frozen=True)
class SystemMessage:
    """System prompt message."""

    content: str
    role: Literal["system"] = "system"


@dataclass(frozen=True)
class UserMessage:
    """User message with multimodal content parts."""

    content: tuple[MessageContent, ...]
    role: Literal["user"] = "user"

    @classmethod
    def from_text(cls, text: str) -> UserMessage:
        """Create a UserMessage with a single TextContent part."""
        return cls(content=(TextContent(text=text),))


@dataclass(frozen=True)
class AssistantMessage:
    """Assistant response message with optional tool calls."""

    content: tuple[MessageContent, ...]
    tool_calls: tuple[ToolCall, ...]
    finish_reason: str | None = None
    role: Literal["assistant"] = "assistant"


@dataclass(frozen=True)
class ToolResultMessage:
    """Tool execution result message."""

    content: str
    tool_call_id: str | None = None
    tool_name: str | None = None
    role: Literal["tool"] = "tool"


Message = SystemMessage | UserMessage | AssistantMessage | ToolResultMessage


# ── Span types ────────────────────────────────────────────────────────────────


@dataclass(frozen=True)
class LLMSpan:
    """One LLM provider call, produced by LLMSpanProcessor from OpenInference attrs."""

    # ── Identity
    span_id: str  # hex, 16 chars
    trace_id: str  # hex, 32 chars
    parent_span_id: str | None  # links to ObserveSpan.span_id when nested

    # ── Timing
    started_at: datetime
    ended_at: datetime
    duration_ms: float

    # ── Provider / model
    operation: str  # "chat" | "embedding"
    provider: str  # "openai" | "anthropic" | "google" | ...
    request_model: str
    response_model: str | None

    # ── Token usage
    input_tokens: int  # default 0
    output_tokens: int  # default 0
    cache_read_tokens: int  # default 0
    cache_creation_tokens: int  # default 0

    # ── Request parameters
    request_temperature: float | None
    request_max_tokens: int | None
    request_top_p: float | None

    # ── Response metadata
    finish_reasons: tuple[str, ...]  # default ()
    response_id: str | None
    output_type: str | None  # "json" | "text" | None
    error_type: str | None

    # ── Content (populated when capture_content=True)
    input_messages: tuple[Message, ...]  # default ()
    output_messages: tuple[AssistantMessage, ...]  # default ()
    tool_definitions: tuple[ToolDefinition, ...]  # always populated when available


@dataclass(frozen=True)
class ObserveSpan:
    """A user-defined instrumented block, produced when a log() block exits."""

    # ── Identity
    span_id: str  # hex, 16 chars
    trace_id: str  # hex, 32 chars
    parent_span_id: str | None

    # ── Timing
    started_at: datetime
    ended_at: datetime
    duration_ms: float

    # ── User-defined fields
    name: str | None  # optional label for the block
    input: Any  # value passed to log(input=...)
    output: Any  # value set via span.set_output(...)
    metadata: dict[str, Any]  # accumulated via span.set_metadata(k, v)
    error: str | None  # exception type if block raised, else None

File: ts/src/instrumentation/spans.ts
/**
 * Data model types for pixie instrumentation spans.
 *
 * All span types use readonly fields to enforce immutability, mirroring
 * the frozen dataclasses in the Python implementation.
 */

// ── Message content types ────────────────────────────────────────────────────

export interface TextContent {
  readonly type: "text";
  readonly text: string;
}

export interface ImageContent {
  readonly type: "image";
  readonly url: string;
  readonly detail: string | null;
}

export type MessageContent = TextContent | ImageContent;

// ── Tool types ───────────────────────────────────────────────────────────────

export interface ToolCall {
  readonly name: string;
  readonly arguments: Record<string, unknown>;
  readonly id: string | null;
}

export interface ToolDefinition {
  readonly name: string;
  readonly description: string | null;
  readonly parameters: Record<string, unknown> | null;
}

// ── Message types ────────────────────────────────────────────────────────────

export interface SystemMessage {
  readonly role: "system";
  readonly content: string;
}

export interface UserMessage {
  readonly role: "user";
  readonly content: readonly MessageContent[];
}

export interface AssistantMessage {
  readonly role: "assistant";
  readonly content: readonly MessageContent[];
  readonly toolCalls: readonly ToolCall[];
  readonly finishReason: string | null;
}

export interface ToolResultMessage {
  readonly role: "tool";
  readonly content: string;
  readonly toolCallId: string | null;
  readonly toolName: string | null;
}

export type Message =
  | SystemMessage
  | UserMessage
  | AssistantMessage
  | ToolResultMessage;

// ── Factory helpers ──────────────────────────────────────────────────────────

export function createTextContent(text: string): TextContent {
  return { type: "text", text };
}

export function createImageContent(
  url: string,
  detail: string | null = null
): ImageContent {
  return { type: "image", url, detail };
}

export function createSystemMessage(content: string): SystemMessage {
  return { role: "system", content };
}

export function createUserMessageFromText(text: string): UserMessage {
  return { role: "user", content: [createTextContent(text)] };
}

export function createUserMessage(
  content: readonly MessageContent[]
): UserMessage {
  return { role: "user", content };
}

export function createAssistantMessage(opts: {
  content: readonly MessageContent[];
  toolCalls: readonly ToolCall[];
  finishReason?: string | null;
}): AssistantMessage {
  return {
    role: "assistant",
    content: opts.content,
    toolCalls: opts.toolCalls,
    finishReason: opts.finishReason ?? null,
  };
}

export function createToolResultMessage(opts: {
  content: string;
  toolCallId?: string | null;
  toolName?: string | null;
}): ToolResultMessage {
  return {
    role: "tool",
    content: opts.content,
    toolCallId: opts.toolCallId ?? null,
    toolName: opts.toolName ?? null,
  };
}

// ── Span types ───────────────────────────────────────────────────────────────

/**
 * One LLM provider call, produced by LLMSpanProcessor from
 * OpenInference attributes.
 */
export interface LLMSpan {
  // Identity
  readonly spanId: string;
  readonly traceId: string;
  readonly parentSpanId: string | null;

  // Timing
  readonly startedAt: Date;
  readonly endedAt: Date;
  readonly durationMs: number;

  // Provider / model
  readonly operation: string;
  readonly provider: string;
  readonly requestModel: string;
  readonly responseModel: string | null;

  // Token usage
  readonly inputTokens: number;
  readonly outputTokens: number;
  readonly cacheReadTokens: number;
  readonly cacheCreationTokens: number;

  // Request parameters
  readonly requestTemperature: number | null;
  readonly requestMaxTokens: number | null;
  readonly requestTopP: number | null;

  // Response metadata
  readonly finishReasons: readonly string[];
  readonly responseId: string | null;
  readonly outputType: string | null;
  readonly errorType: string | null;

  // Content
  readonly inputMessages: readonly Message[];
  readonly outputMessages: readonly AssistantMessage[];
  readonly toolDefinitions: readonly ToolDefinition[];
}

/**
 * A user-defined instrumented block, produced when a
 * startObservation() block exits.
 */
export interface ObserveSpan {
  // Identity
  readonly spanId: string;
  readonly traceId: string;
  readonly parentSpanId: string | null;

  // Timing
  readonly startedAt: Date;
  readonly endedAt: Date;
  readonly durationMs: number;

  // User-defined fields
  readonly name: string | null;
  readonly input: unknown;
  readonly output: unknown;
  readonly metadata: Readonly<Record<string, unknown>>;
  readonly error: string | null;
}
